vnext/Desktop.IntegrationTests/HttpResourceIntegrationTests.cpp

```diff
@@ -354,6 +354,104 @@ TEST_CLASS (HttpResourceIntegrationTest) {
+
+    promise<void> getPromise;
+    string error;
+    int statusCode = 0;
+
+    auto server = std::make_shared<HttpServer>(s_port);
+    server->Callbacks().OnGet = [&getPromise](const DynamicRequest &request) -> ResponseWrapper {
+      DynamicResponse response;
+      response.result(http::status::ok);
+
+      // Hold response to test client timeout
+      promise<void> timer;
+      timer.get_future().wait_for(std::chrono::milliseconds(2000));
+
+      return {std::move(response)};
+    };
+    server->Start();
+
+    auto resource = IHttpResource::Make();
+    resource->SetOnResponse([&getPromise, &statusCode](int64_t, IHttpResource::Response response) {
+      statusCode = static_cast<int>(response.StatusCode);
+      getPromise.set_value();
+    });
+    resource->SetOnError([&getPromise, &error, &server](int64_t, string &&errorMessage) {
+      error = std::move(errorMessage);
+      getPromise.set_value();
+    });
+    resource->SendRequest(
+        "GET",
+        std::move(url),
+        0, /*requestId*/
+        {}, /*headers*/
+        {}, /*data*/
+        "text", /*responseType*/
+        false, /*useIncrementalUpdates*/
+        6000, /*timeout*/
+        false, /*withCredentials*/
+        [](int64_t) {} /*callback*/);
+
+    getPromise.get_future().wait();
+    server->Stop();
+
+    Assert::AreEqual({}, error);
+    Assert::AreEqual(200, statusCode);
+  }
+
+  TEST_METHOD(TimeoutFails) {
+    auto port = s_port;
+    string url = "http://localhost:" + std::to_string(port);
+
+    promise<void> getPromise;
+    string error;
+    int statusCode = 0;
+
+    auto server = std::make_shared<HttpServer>(s_port);
+    server->Callbacks().OnGet = [&getPromise](const DynamicRequest &request) -> ResponseWrapper {
+      DynamicResponse response;
+      response.result(http::status::ok);
+
+      // Hold response to test client timeout
+      promise<void> timer;
+      timer.get_future().wait_for(std::chrono::milliseconds(4000));
+
+      return {std::move(response)};
+    };
+    server->Start();
+
+    auto resource = IHttpResource::Make();
+    resource->SetOnResponse([&getPromise, &statusCode](int64_t, IHttpResource::Response response) {
+      statusCode = static_cast<int>(response.StatusCode);
+      getPromise.set_value();
+    });
+    resource->SetOnError([&getPromise, &error, &server](int64_t, string &&errorMessage) {
+      error = std::move(errorMessage);
+      getPromise.set_value();
+    });
+    resource->SendRequest(
+        "GET",
+        std::move(url),
+        0, /*requestId*/
+        {}, /*headers*/
+        {}, /*data*/
+        "text", /*responseType*/
+        false, /*useIncrementalUpdates*/
+        2000, /*timeout*/
+        false, /*withCredentials*/
+        [](int64_t) {} /*callback*/);
+
+    getPromise.get_future().wait();
+    server->Stop();
+
+    Assert::AreEqual({"[0x800705b4] This operation returned because the timeout period expired."}, error);
+    Assert::AreEqual(0, statusCode);
```

Comment:
>0, statusCode [](http://example.com/codeflow?start=21&length=13)

I don't fully understand the meaning of statusCode here, but shouldn't there be a 408 in here somewhere ([https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408))?

Reply:
Good question.

Not quite.

This change is to implement client-side timeout.
408 is a server-side timeout code, which means there was a response from the server in the first place.

In this situation, we never get a response and no thus no HTTP error code to deal with.

Reply:
To further clarify:

Server-side timeout: Server determines the request has taken too much time, bails and notifies the client.

Client-side timeout (this PR): Client determines the request has taken too long to complete (or even begin), and bails.

Reply:
I see. What's the way the client can figure out that a request timed out?

Reply:
This is actually an arbitrary parameter passed from the JavaScript layer.
The RN client decides what's a tolerable timeout threshold (0 being no timeout).

Reply:
Right, but how does client know that a request timed out?